src/doc/design_overview/overview.rst

```diff
@@ -0,0 +1,961 @@
+keep individual clients connected to the same hub instance).
+
+Hub
+^^^
+
+The "Hub" is the primary server backend for SMC, built on Node.js as described
+previously.  It consists of an HTTP server with Primus + Engine.IO attached to
+handle real-time bidirectional client/hub communication.  Most communication
+between the Client and the backend happens through the Hub, whose HTTP server
+uses Express to route requests to different services (account management,
+project management, payment, etc.).  Each Hub also sets up a
+``ComputeServerClient`` which gives it access to all the running compute
+servers (discussed next).  The names and URLs of all the available compute
+servers live in a system table in the database.
+
+It also uses `node-http-proxy`_ to create an HTTP proxy server associated
+with each Hub (on port number one higher than the Hub's HTTP port).  If we
+understand correctly, the proxy handles all requests that are to be
+forwarded to individual compute nodes (such as requesting files, or
+resources on web servers belonging to a specific project).  HAproxy doesn't
+know anything about the compute nodes themseves--it just sees URLs that look
+like they belong to a project (they begin with a project UUID) and forwards
+those requests to the Hub's proxy, which in turn checks that the requester
+is authenticated and has permissions to access that project's resources.
+The proxy then forwards the request to the appropriate port on that
+project's compute node::
+
+    Client <--> HAProxy <--> Hub Proxy <--> Compute
+
+Compute
+^^^^^^^
+
+Compute servers are where the real work gets done in SMC projects.  Every
+project is associated with a specific compute server where all their data
+is stored (by way of storage servers mounted on the compute node) and where
+all process and computation tasks done by the project are performed.  This
+includes running Sage.  The Compute servers are Linux VMs with varying
+degrees of hardware capacity, depending on how much you're willing to pay.
+In most cases the servers are shared between projects (you don't have admin
+on the servers) though in principle one could pay for one's own compute
+server as well.
+
+Otherwise, one can do quite a bit of different things on their compute node,
+including log in to the shell directly (you log in as a user named after
+your project's UUID).  This can be done either through the web terminal in
+SMC, or one can SSH in directly.
+
+Each compute node also runs a simple socket server
+``smc-hub/compute-server.coffee`` that is used by the Hub to communicate with
+the compute node (using simple JSON messages). For example, one can make status
+inquiries on the node, or send commands to run a command in a project.
+
+Additionally, each *project* runs a couple per-project daemons when the
+project is created and running.  These include the the console server
+(``smc-project/console_server.coffee``) which provides the backend for the
```

Comment:
Originally -- for efficiency reasons (both coding and runtime) -- it was because I didn't want to use the same TCP connection for a terminal as for everything else (all the JSON control messages).  However, I just bit the bullet and added this extra support...